Sources/MapboxDirections/DirectionsOptions.swift

```diff
@@ -369,6 +395,96 @@ open class DirectionsOptions: Codable {
+            guard let value = $1.value else { return $0 }
+            return $0.merging([$1.name: value]) { (current, _) in current }
+        }) else {
+            return
+        }
+        
+        if let mappedValue = mappedQueryItems[CodingKeys.shapeFormat.stringValue],
+           let shapeFormat = RouteShapeFormat(rawValue: mappedValue) {
+            self.shapeFormat = shapeFormat
+        }
+        if let mappedValue = mappedQueryItems[CodingKeys.routeShapeResolution.stringValue],
+           let routeShapeResolution = RouteShapeResolution(rawValue: mappedValue) {
+            self.routeShapeResolution = routeShapeResolution
+        }
+        if mappedQueryItems[CodingKeys.includesSteps.stringValue] == "true" {
+            self.includesSteps = true
+        }
+        if let mappedValue = mappedQueryItems[CodingKeys.locale.stringValue] {
+            self.locale = Locale(identifier: mappedValue)
+        }
+        if mappedQueryItems[CodingKeys.includesSpokenInstructions.stringValue] == "true" {
+            self.includesSpokenInstructions = true
+        }
+        if let mappedValue = mappedQueryItems[CodingKeys.distanceMeasurementSystem.stringValue],
+           let measurementSystem = MeasurementSystem(rawValue: mappedValue) {
+            self.distanceMeasurementSystem = measurementSystem
+        }
+        if mappedQueryItems[CodingKeys.includesVisualInstructions.stringValue] == "true" {
+            self.includesVisualInstructions = true
+        }
+        if let mappedValue = mappedQueryItems[CodingKeys.attributeOptions.stringValue],
+           let attributeOptions = AttributeOptions(descriptions: mappedValue.components(separatedBy: ",")) {
+            self.attributeOptions = attributeOptions
+        }
+        if let mappedValue = mappedQueryItems["waypoints"] {
+            let indicies = mappedValue.components(separatedBy: ";").compactMap { Int($0) }
+            if !indicies.isEmpty {
+                waypoints.enumerated().forEach {
+                    $0.element.separatesLegs = indicies.contains($0.offset)
+                }
+            }
+        }
+        
+        let waypointsData = [mappedQueryItems["approaches"]?.components(separatedBy: ";"),
+                             mappedQueryItems["bearings"]?.components(separatedBy: ";"),
+                             mappedQueryItems["radiuses"]?.components(separatedBy: ";"),
+                             mappedQueryItems["waypoint_names"]?.components(separatedBy: ";"),
+                             mappedQueryItems["snapping_include_closures"]?.components(separatedBy: ";")
+        ] as [[String]?]
+        
+        let getElement: ((_ array: [String]?, _ index: Int) -> String?) = { array, index in
+            if array?.count ?? -1 > index {
+                return array?[index]
+            }
+            return nil
+        }
+        
+        waypoints.enumerated().forEach {
+            if let approach = getElement(waypointsData[0], $0.offset) {
```

Comment:
using `zip`s was my previous approach. It results in multiple cycles iterating over `waypoints` for each parameter. This one at least is easier to understand and see that it's all the properties that are applied to waypoints parameters. Probably there could be a method to decode a waypoints array from an URL for clarity, but I am not sure it's worth it.

Reply:
The main issue here is that using hardcoded indices is a dangerous source of bugs. Do you have any ideas on how to avoid this? I can imagine use of some enum here, not sure

Reply:
I don't think something like enum would make sense here because `waypointsData` is still filled manually here. Ideally I would factor `waypointsData` into special wrapper which does this data manipulations, but I think it is an overkill for the sake of this method.